lib/Dialect/FIRRTL/Transforms/InferResets.cpp

```diff
@@ -0,0 +1,1538 @@
+static std::pair<StringAttr, FModuleOp> getResetNameAndModule(Value reset) {
+  if (auto arg = reset.dyn_cast<BlockArgument>()) {
+    auto module = cast<FModuleOp>(arg.getParentRegion()->getParentOp());
+    return {getModulePortName(module, arg.getArgNumber()), module};
+  } else {
+    auto op = reset.getDefiningOp();
+    return {op->getAttrOfType<StringAttr>("name"),
+            op->getParentOfType<FModuleOp>()};
+  }
+}
+
+/// Return the name of a reset. The reset value must either be a module port or
+/// a wire/node operation.
+static inline StringAttr getResetName(Value reset) {
+  return getResetNameAndModule(reset).first;
+}
+
+/// Construct a zero value of the given type using the given builder.
+static Value createZeroValue(ImplicitLocOpBuilder &builder, FIRRTLType type,
+                             SmallDenseMap<FIRRTLType, Value> &cache) {
+  auto it = cache.find(type);
+  if (it != cache.end())
+    return it->second;
+  auto nullBit = [&]() {
+    return createZeroValue(builder, UIntType::get(builder.getContext(), 1),
+                           cache);
+  };
+  auto value =
+      TypeSwitch<FIRRTLType, Value>(type)
+          .Case<ClockType>([&](auto type) {
+            return builder.create<AsClockPrimOp>(nullBit());
+          })
+          .Case<AsyncResetType>([&](auto type) {
+            return builder.create<AsAsyncResetPrimOp>(nullBit());
+          })
+          .Case<SIntType, UIntType>([&](auto type) {
+            return builder.create<ConstantOp>(
+                type, APInt::getNullValue(type.getWidth().getValueOr(1)));
+          })
+          .Case<BundleType>([&](auto type) {
+            auto wireOp = builder.create<WireOp>(type);
+            for (auto &field : type.getElements()) {
+              auto zero = createZeroValue(builder, field.type, cache);
+              auto acc =
+                  builder.create<SubfieldOp>(field.type, wireOp, field.name);
+              builder.create<ConnectOp>(acc, zero);
+            }
+            return wireOp;
+          })
+          .Case<FVectorType>([&](auto type) {
+            auto wireOp = builder.create<WireOp>(type);
+            auto zero = createZeroValue(builder, type.getElementType(), cache);
+            for (unsigned i = 0, e = type.getNumElements(); i < e; ++i) {
+              auto acc = builder.create<SubindexOp>(zero.getType(), wireOp, i);
+              builder.create<ConnectOp>(acc, zero);
+            }
+            return wireOp;
+          })
+          .Case<ResetType, AnalogType>(
+              [&](auto type) { return builder.create<InvalidValueOp>(type); })
+          .Default([](auto) {
+            llvm_unreachable("switch handles all types");
+            return Value{};
+          });
+  cache.insert({type, value});
+  return value;
+}
+
+/// Construct a null value of the given type using the given builder.
+static Value createZeroValue(ImplicitLocOpBuilder &builder, FIRRTLType type) {
+  SmallDenseMap<FIRRTLType, Value> cache;
```

Comment:
Did you mean this cache to be persistent?  It's destroyed at function exit.  It also isn't tied to the context in anyway, so in theory to runs of this pass with different contexts could mix operations.

Reply:
This is intended to just live for the duration of the `createZeroValue` call. Since the `createZeroValue` assembles potentially large bundles and vectors where types are likely to repeat, this cache allows the function to only construct one zero value per type. On separate calls you still get separate constants (to be picked up by CSE).